components/eamxx/src/physics/p3/impl/p3_init_impl.hpp

```diff
@@ -3,35 +3,327 @@
+      dum5 = 0; // term for ventilation factor in evap
+      dd   = 2;
+
+      // loop over PSD to numerically integrate number and mass-weighted mean fallspeeds
+      for (kk = 1; kk <= 10000; ++kk) {
+
+        dia = (kk*dd - dd*0.5)*1.e-6;     // size bin [m]
+        amg = c::PIOV6*997 * std::pow(dia, 3); // mass [kg]
+        amg = amg*1000;                   // convert [kg] to [g]
+
+        // get fallspeed as a function of size [m s-1]
+        if (dia*1.e+6 <= 134.43) {
+          vt = 4.5795e+3 * std::pow(amg, 2*thrd);
+        }
+        else if (dia*1.e+6 < 1511.64) {
+          vt = 4.962e+1 * std::pow(amg, thrd);
+        }
+        else if (dia*1.e+6 < 3477.84) {
+          vt = 1.732e+1 * std::pow(amg, c::SXTH);
+        }
+        else {
+          vt = 9.17;
+        }
+
+        // note: factor of 4.*mu_r is non-answer changing and only needed to
+        //       prevent underflow/overflow errors, same with 3.*mu_r for dum5
+        dum1 += vt * std::pow(10, mu_r*std::log10(dia) + 4*mu_r) * std::exp(-lamr*dia) * dd * 1.e-6;
+        dum2 += std::pow(10, mu_r*std::log10(dia) + 4*mu_r) * std::exp(-lamr*dia) * dd * 1.e-6;
+        dum3 += vt * std::pow(10, (mu_r+3)*std::log10(dia) + 4*mu_r) * std::exp(-lamr*dia) * dd * 1.e-6;
+        dum4 += std::pow(10, (mu_r+3)*std::log10(dia) + 4*mu_r) * std::exp(-lamr*dia) * dd * 1.e-6;
+        dum5 += std::pow(vt*dia, 0.5) * std::pow(10, (mu_r+1)*std::log10(dia) + 3*mu_r) * std::exp(-lamr*dia) * dd * 1.e-6;
+      }
+
+      dum2 = std::max(dum2, 1.e-30); // to prevent divide-by-zero below
+      dum4 = std::max(dum4, 1.e-30); // to prevent divide-by-zero below
+      dum5 = std::max(dum5, 1.e-30); // to prevent log10-of-zero below
+
+      vn_table_vals_nc(jj-1,ii-1)    = dum1/dum2;
+      vm_table_vals_nc(jj-1,ii-1)    = dum3/dum4;
+      revap_table_vals_nc(jj-1,ii-1) = std::pow(10, std::log10(dum5) + (mu_r+1)*std::log10(lamr) - (3*mu_r));
+    }
+  }
+
+  mu_r_table_vals = mu_r_table_vals_nc;
+  vn_table_vals = vn_table_vals_nc;
+  vm_table_vals = vm_table_vals_nc;
+  revap_table_vals = revap_table_vals_nc;
+}
+
+template <bool IsRead>
+struct IoAction
+{
+  template <typename S>
+  static void action(const ekat::FILEPtr& fid, S* data, const size_t size)
+  {
+    ekat::read(data, size, fid);
+  }
+};
+
+template <>
+struct IoAction<false>
+{
+  template <typename S>
+  static void action(const ekat::FILEPtr& fid, S* data, const size_t size)
+  {
+    ekat::write(data, size, fid);
+  }
+};
+
+template <bool IsRead, typename MuRT, typename VNT, typename VMT, typename RevapT>
+void io_impl(const bool masterproc, const char* dir, MuRT& mu_r_table_vals, VNT& vn_table_vals, VMT& vm_table_vals, RevapT& revap_table_vals)
+{
+  using Action = IoAction<IsRead>;
+
+  if (masterproc) {
+    std::cout << (IsRead ? "Reading" : "Writing") << " lookup (non-ice) tables in dir " << dir << std::endl;
+  }
+
+  std::string extension =
+#ifdef SCREAM_DOUBLE_PRECISION
```

Comment:
Idea: since we are in the middle of rearranging our tables, why not being more verbose with names? Maybe instead of appending `8` or `4`, we could append the actual type name? Like `mu_r_table_vals_double.dat`?

Yes, it will require changing defaults in the xml, but if clarity increases, it's worth it, imho.

Reply:
This is the convention we are using all over, for better or worse.